### PR TITLE
Asio Dependency [5404]

### PR DESCRIPTION
**Pull request by julianbermudez**
This pull request removes the Asio dependency.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/173/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/176/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/181/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/184/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/182/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/185/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/196/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/199/)

File: include/uxr/agent/utils/TokenBucket.hpp [hunk 1]
```diff
@@ -25,17 +26,24 @@ namespace utils {
...
-    explicit TokenBucket(size_t rate, size_t burst = min_rate_);
-    TokenBucket(TokenBucket&& other) noexcept;
+    explicit TokenBucket(
+            size_t rate,
+            size_t burst = min_rate_);
+
+    ~TokenBucket() = default;
+
+    TokenBucket(TokenBucket&& other) = default;
     TokenBucket(const TokenBucket& other);
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/utils/TokenBucket.hpp]: Removed the custom operator and set by default.

File: include/uxr/agent/utils/TokenBucket.hpp [hunk 2]
```diff
@@ -44,8 +52,97 @@ class TokenBucket
...
+    auto current_time = std::chrono::steady_clock::now();
+    uint64_t delta_time =
+            uint64_t(std::chrono::duration_cast<std::chrono::milliseconds>(current_time - timestamp_).count());
+    tokens_ = std::min(capacity_, tokens_ + size_t((rate_ * delta_time) / std::milli::den));
+    timestamp_ = current_time;
+}
+
+inline size_t TokenBucket::get_tokens_available()
+{
+    auto current_time = std::chrono::steady_clock::now();
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/utils/TokenBucket.hpp]: Refectorized ;).

File: src/cpp/datareader/DataReader.cpp [hunk 3]
```diff
@@ -26,6 +26,12 @@
 namespace eprosima {
 namespace uxr {
 
+const int READ_TIMEOUT = 100;
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/datareader/DataReader.cpp]: Yeah, good idea!

File: src/cpp/datareader/DataReader.cpp [hunk 4]
```diff
@@ -150,46 +156,60 @@ bool DataReader::stop_read()
...
     uint16_t message_count = 0;
+    std::chrono::time_point<std::chrono::steady_clock> init_time;
+    init_time = std::chrono::steady_clock::now();
 
-    while (running_cond_ && (message_count < delivery_control.max_samples()))
+    while (running_cond_ && !stop_cond)
     {
         std::vector<uint8_t> data;
-        if (get_middleware().read_data(get_raw_id(), data, std::chrono::milliseconds(100)))
+        if (get_middleware().read_data(get_raw_id(), data, std::chrono::milliseconds(READ_TIMEOUT)))
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/datareader/DataReader.cpp]: Fixed!

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/205/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/208/)

**Pull request merged**